packages/block-library/src/site-logo/index.php

```diff
@@ -111,54 +111,34 @@ function _override_custom_logo_theme_mod( $custom_logo ) {
 /**
  * Updates the site_logo option when the custom_logo theme-mod gets updated.
  *
- * This function is hooked on "update_option_theme_mods_$theme" and not
- * "pre_set_theme_mod_custom_logo" because by hooking in `update_option`
- * the function accounts for remove_theme_mod() as well.
+ * This function is hooked on "pre_update_option_theme_mods_$theme" and not
+ * "pre_set_theme_mod_custom_logo" because by hooking into the theme_mods
+ * option the function accounts for remove_theme_mod() as well.
  *
- * @param mixed $old_value The old option value.
- * @param mixed $value     The new option value.
+ * The `pre_update_` hook is used so the `site_logo` option will be updated even
+ * if the new custom_logo value is the same as the existing one, or the option
+ * does not exist yet. In both cases the regular `update_` hook is not called.
+ *
+ * @param array $value Theme mod settings.
+ * @return array
  */
-function _sync_custom_logo_to_site_logo( $old_value, $value ) {
-	// Delete the option when the custom logo does not exist or was removed.
-	// This step ensures the option stays in sync.
+function _sync_custom_logo_to_site_logo( $value ) {
 	if ( empty( $value['custom_logo'] ) ) {
 		delete_option( 'site_logo' );
 	} else {
-		remove_action( 'update_option_site_logo', '_sync_site_logo_to_custom_logo' );
 		update_option( 'site_logo', $value['custom_logo'] );
-		add_action( 'update_option_site_logo', '_sync_site_logo_to_custom_logo', 10, 2 );
 	}
+
+	return $value;
 }
 
 /**
- * Hooks `_sync_custom_logo_to_site_logo` in `update_option_theme_mods_$theme`.
+ * Hooks `_sync_custom_logo_to_site_logo` into `pre_update_option_theme_mods_$theme`.
  *
  * Runs on `setup_theme` to account for dynamically-switched themes in the Customizer.
  */
 function _sync_custom_logo_to_site_logo_on_setup_theme() {
 	$theme = get_option( 'stylesheet' );
-	add_action( "update_option_theme_mods_$theme", '_sync_custom_logo_to_site_logo', 10, 2 );
+	add_filter( "pre_update_option_theme_mods_$theme", '_sync_custom_logo_to_site_logo', 10 );
 }
 add_action( 'setup_theme', '_sync_custom_logo_to_site_logo_on_setup_theme', 11 );
-
-/**
- * Updates the custom_logo theme-mod when the site_logo option gets updated.
- *
- * @param mixed $old_value The old option value.
- * @param mixed $value     The new option value.
- *
- * @return void
- */
-function _sync_site_logo_to_custom_logo( $old_value, $value ) {
-	// Delete the option when the custom logo does not exist or was removed.
-	// This step ensures the option stays in sync.
-	if ( empty( $value ) ) {
-		remove_theme_mod( 'custom_logo' );
-	} else {
-		remove_filter( 'pre_set_theme_mod_custom_logo', '_sync_custom_logo_to_site_logo' );
-		set_theme_mod( 'custom_logo', $value );
-		add_filter( 'pre_set_theme_mod_custom_logo', '_sync_custom_logo_to_site_logo' );
```

Comment:
Note that this line has a bug in it that can erase all theme mods by calling `_sync_custom_logo_to_site_logo` with just the custom logo value instead of with the theme mods array. If we don't remove this code, we at least need to fix this filter.